File: 2022/easy/first-bad-version.cpp
//  https://leetcode.com/problems/first-bad-version/

#include <iostream>
// The API isBadVersion is defined for you.
bool isBadVersion(int version) {
    return version > 3;
}

class Solution {
public:
    int firstBadVersion(int n) {
        int left, right, mid;
        if (n == 1) return isBadVersion(n);
        left = 1; 
        right = n;

    
        while (left <= right) {
            mid = left + (right - left) / 2;

            bool midIsBad = isBadVersion(mid);
    
            if (midIsBad && (mid - 1 == 0 || !isBadVersion(mid - 1))) {
                return mid;
            }
            if (midIsBad) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return -1;
    }
};

int main () {
    auto s = new Solution();
    std::cout << s->firstBadVersion(20);
    return 0;
}

File: 2022/grind75/first-bad-version.ts
// https://leetcode.com/problems/first-bad-version/

var solution = function(isBadVersion: any) {

    return function(n: number): number {
        let left = 0;
        let right = n;

        while (left <= right) {
            let mid = Math.trunc((left + right) / 2)
            const isBad = isBadVersion(mid)
            if (left == right) {
                if (isBad) return left;
                return left + 1 
            }
            if (!isBad) {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }

        return left
    };
};

let sol = solution((n: number)=>n >= 2)

console.log(sol(3))